src/core/lib/event_engine/forkable.h

```diff
@@ -34,41 +37,34 @@ extern grpc_core::TraceFlag grpc_trace_fork;
 
 #define GRPC_FORK_TRACE_LOG_STRING(format) GRPC_FORK_TRACE_LOG("%s", format)
 
-// Register fork handlers with the system, enabling fork support.
-//
-// This provides pthread-based support for fork events. Any objects that
-// implement Forkable can register themselves with this system using
-// ManageForkable, and their respective methods will be called upon fork.
-//
-// This should be called once upon grpc_initialization.
-void RegisterForkHandlers();
-
-// Global callback for pthread_atfork's *prepare argument
-void PrepareFork();
-// Global callback for pthread_atfork's *parent argument
-void PostforkParent();
-// Global callback for pthread_atfork's *child argument
-void PostforkChild();
-
 // An interface to be implemented by EventEngines that wish to have managed fork
 // support.
 class Forkable {
  public:
-  Forkable();
-  virtual ~Forkable();
+  virtual ~Forkable() = default;
   virtual void PrepareFork() = 0;
   virtual void PostforkParent() = 0;
   virtual void PostforkChild() = 0;
 };
 
-// Add Forkables from the set of objects that are supported.
-// Upon fork, each forkable will have its respective fork hooks called on
-// the thread that invoked the fork.
-//
-// Relative ordering of fork callback operations is not guaranteed.
-void ManageForkable(Forkable* forkable);
-// Remove a forkable from the managed set.
-void StopManagingForkable(Forkable* forkable);
+#ifndef GRPC_POSIX_FORK_ALLOW_PTHREAD_ATFORK
+// no-op ObjectGroupForkHandler impl
+#else
+class ObjectGroupForkHandler {
+ public:
+  // adds forkable to set of forkables
+  // asserts if IsForking() == true
+  void RegisterForkable(std::shared_ptr<Forkable> forkable);
+
+  void Prefork();
+  void PostforkParent();
+  void PostforkChild();
+
+ private:
+  bool is_forking_ = false;
+  std::vector<std::weak_ptr<Forkable> > forkables_;
```

Comment:
How does `forkables_` shrink?

Reply:
Good point. I added logic to shrink `forkables_` when the object has been destroyed in https://github.com/grpc/grpc/pull/33733/commits/2fdb0af96de913bfcf15c31f1eed7e83dfc0429b. PTAL.

Reply:
We may have to invest in shrinking the vector of weak_ptrs for the majority of times when a `fork` never occurs. It may be ok to make that a future problem, given the small set of usually-long-lived objects being tracked here.